source/docs/software/vision-processing/apriltag/apriltag-intro.rst

```diff
@@ -0,0 +1,233 @@
+      * Use these gradient measurements to rapidly re-fit an exterior quadrilateral at full resolution
+      * Use geometry to calculate the exact center of the re-fit quadrilateral
+
+      Note that this step is optional, and can be skipped for faster image processing. However, skipping it can induce significant errors into your robot's behavior, depending on how you are using the tag outputs.
+
+
+Usage
+-----
+
+2D Alignment
+^^^^^^^^^^^^
+
+A simple strategy for using targets is to move the robot until the target is centered in the image. Assuming the field and robot are constructed such that the gamepiece, scoring location, vision target, and camera are all aligned, this method should proved a straightforward method to automatically align the robot to the scoring position.
+
+Using a camera, identify the _centroid_ of the apriltags in view. If the tag's ID is correct, apply drivetrain commands to rotate the robot left or right until the tag is centered in the camera image.
+
+This method does not require calibrating the camera or performing the homography step.
+
+.. tabs::
+
+  .. code-tab:: java
+
+    // Coming soon!
+
+  .. code-tab:: c++
+
+    // Coming soon!
+
+3D Alignment
+^^^^^^^^^^^^
+
+.. image:: images/homography.png
+
+A more advanced usage of AprilTags is to use their corner locations to help perform on-field localization.
+
+This method requires calibrating the camera to measure how its optics distort 3d space onto the 2d sensor.
+
+Once calibrated, each image is searched for AprilTags using the above algorithm.
+
+Given each tag's ID, the position of the tag on the field can be identified.
+
+.. tabs::
+
+  .. code-tab:: java
+
+    // Coming soon!
+
+  .. code-tab:: c++
+
+    // Coming soon!
+
+
+Using the information about the camera's distortion, along with the known size of the tag, an estimate of the camera's position relative to the tag is calculated.
+
+.. tabs::
+
+  .. code-tab:: java
+
+    // Coming soon!
+
+  .. code-tab:: c++
+
+    // Coming soon!
+
+In turn, using the `Pose3d` classes and the known positions of tags on the field, the robot's position on the field may be estimate.
+
+These estimates can be incorporated into the WPILib pose estimation classes.
+
+2D to 3D Ambiguity
+------------------
+
+The process of translating the four known corners of the target in the image (two-dimensional) into a real-world position relative to the camera (three-dimensional") is inherently ambiguous. That is to say, there are multiple real-world positions that result in the target corners ending up in the same spot in the image.
+
+Humans can often use lighting or background clues to understand how objects are oriented in space. However, computers do not have this benefit. They can be tricked by similar-looking targets:
+
+.. image:: images/planar_ambiguity1.png
+
+.. image:: images/planar_ambiguity2.png
+
+.. image:: images/planar_ambiguity3.png
```

Comment:
I think there's a directive to put images side by side, or can we use a photo editor and combine these images?